### PR TITLE
62 support assumptions

### LINKED ISSUE #62: Support assumptions
Example of syntax we would like to be able to use:
```java
assumeThat(someValidation).isValid();
```

i.e. support equivalent syntax for assumptions.

Assumptions are particularly useful when used in combination with Property Based Testing (PBT), e.g. with JUnit-QuickCheck.

### PR DESCRIPTION
**Pull request by mchmielarz**
Fixes #62

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -81,12 +99,29 @@
...
         </dependency>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <scope>provided</scope>
+            <optional>true</optional>
+        </dependency>
     </dependencies>
 
     <build>
```
Comment (Opalo) [inline pom.xml]: I'm not sure how the dependencies are resolved in maven but wouldn't it be safer to declare the along with versions?

File: src/main/java/org/assertj/vavr/api/AbstractEitherAssert.java [hunk 2]
```diff
@@ -27,10 +26,7 @@
 import static org.assertj.core.util.Preconditions.checkArgument;
 import static org.assertj.vavr.api.EitherShouldBeLeft.shouldBeLeft;
 import static org.assertj.vavr.api.EitherShouldBeRight.shouldBeRight;
-import static org.assertj.vavr.api.EitherShouldContain.shouldContainOnLeft;
-import static org.assertj.vavr.api.EitherShouldContain.shouldContainOnRight;
-import static org.assertj.vavr.api.EitherShouldContain.shouldContainSameOnLeft;
-import static org.assertj.vavr.api.EitherShouldContain.shouldContainSameOnRight;
+import static org.assertj.vavr.api.EitherShouldContain.*;
```
Comment (Opalo) [inline src/main/java/org/assertj/vavr/api/AbstractEitherAssert.java]: 👍

File: src/main/java/org/assertj/vavr/api/AbstractTryAssert.java [hunk 3]
```diff
@@ -14,14 +14,9 @@
...
 import io.vavr.control.Try;
-import org.assertj.core.api.AbstractAssert;
 import org.assertj.core.api.Assertions;
 import org.assertj.core.api.Condition;
-import org.assertj.core.internal.ComparatorBasedComparisonStrategy;
-import org.assertj.core.internal.ComparisonStrategy;
-import org.assertj.core.internal.Conditions;
-import org.assertj.core.internal.FieldByFieldComparator;
-import org.assertj.core.internal.StandardComparisonStrategy;
+import org.assertj.core.internal.*;
```
Comment (Opalo) [inline src/main/java/org/assertj/vavr/api/AbstractTryAssert.java]: 👍

File: src/main/java/org/assertj/vavr/api/ClassLoadingStrategyFactory.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+    static {
+        Method privateLookupIn;
+        try {
+            privateLookupIn = MethodHandles.class.getMethod("privateLookupIn", Class.class, MethodHandles.Lookup.class);
+        } catch (Exception e) {
+            privateLookupIn = null;
+        }
+        PRIVATE_LOOKUP_IN = privateLookupIn;
+    }
```
Comment (Opalo) [inline src/main/java/org/assertj/vavr/api/ClassLoadingStrategyFactory.java]: Why don't you use vavr here?

File: src/main/java/org/assertj/vavr/api/ClassLoadingStrategyFactory.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+    static ClassLoadingStrategy<ClassLoader> classLoadingStrategy(Class<?> assertClass) {
+        if (ClassInjector.UsingReflection.isAvailable()) {
+            return ClassLoadingStrategy.Default.INJECTION;
+        } else if (ClassInjector.UsingLookup.isAvailable()) {
+            try {
+                return ClassLoadingStrategy.UsingLookup.of(PRIVATE_LOOKUP_IN.invoke(null, assertClass, LOOKUP));
+            } catch (Exception e) {
+                throw new IllegalStateException("Could not access package of " + assertClass, e);
```
Comment (Opalo) [inline src/main/java/org/assertj/vavr/api/ClassLoadingStrategyFactory.java]: Here `PRIVATE_LOOKUP_IN` potentially might be null, so this error might be misleading.

File: src/main/java/org/assertj/vavr/api/VavrAssumptions.java [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+     * @param actual  the actual value.
+     * @return the created assumption for assertion object.
+     */
+    @CheckReturnValue
+    @SuppressWarnings("unchecked")
+    public static <INVALID, VALID> AbstractValidationAssert<?, INVALID, VALID> assumeThat(Validation<INVALID, VALID> actual) {
+        return asAssumption(ValidationAssert.class, Validation.class, actual);
+    }
+
+    // private methods
```
Comment (Opalo) [inline src/main/java/org/assertj/vavr/api/VavrAssumptions.java]: This comment seems to be redundant.

File: src/main/java/org/assertj/vavr/api/VavrAssumptions.java [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+
+    private static <ASSERTION> ASSERTION asAssumption(Class<ASSERTION> assertionType,
+                                                      Class<?>[] constructorTypes,
+                                                      Object... constructorParams) {
+        try {
+            Class<? extends ASSERTION> type = createAssumptionClass(assertionType);
+            Constructor<? extends ASSERTION> constructor = type.getConstructor(constructorTypes);
+            return constructor.newInstance(constructorParams);
+        } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException | InstantiationException e) {
+            throw new RuntimeException(e);
```
Comment (Opalo) [inline src/main/java/org/assertj/vavr/api/VavrAssumptions.java]: Maybe add some `message` here?

File: src/main/java/org/assertj/vavr/api/ClassLoadingStrategyFactory.java (hunk 4 above)
Comment (mchmielarz) [reply src/main/java/org/assertj/vavr/api/ClassLoadingStrategyFactory.java]: Copy-paste from assertj-core ;)

File: src/main/java/org/assertj/vavr/api/ClassLoadingStrategyFactory.java (hunk 5 above)
Comment (mchmielarz) [reply src/main/java/org/assertj/vavr/api/ClassLoadingStrategyFactory.java]: Good spot

File: pom.xml (hunk 1 above)
Comment (mchmielarz) [reply pom.xml]: Maven takes version numbers from `dependencyManagement` section. I think it's totally ok to have the management of versions in this section at least for junit dependencies.

**Pull request merged**